addons/website_sale/models/sale_order.py

```diff
@@ -409,3 +409,41 @@ def _get_shop_warning(self, clear=True):
         if clear:
             self.shop_warning = ''
         return warn
+
+    def _can_send_abandoned_cart_mail(self, sales):
+        self.ensure_one()
+        partner = self.partner_id
+        # Customer needs to be signed in otherwise the mail address is not known.
+        if not partner.email:
+            return False
+        # If a potential customer creates one or more abandoned checkouts and then completes a sale before
+        # the recovery email gets sent, then the email won't be sent.
+        partner_sale_orders = sales.filtered_domain([
```

Comment:
While I am happy we do not query the sales for each one, `filtered_domain` is still unnecessarily complex for our use case.
We could group the sales by partner like so:
```py
related_sales_per_partner = defaultdict(lambda: self.env['sale.order'])
for sale in sales:
    related_sales_per_partner[sale.partner_id] |= sale
```
and pass it to the function.

Also I am not a big fan of `sales` being provided to the function instead this function being able to take a recordset and returning a subset of that recordset.